pkg/test/common.go

```diff
@@ -167,6 +167,43 @@ func CopyFiles(sourceDir, destDir string) error {
 	return nil
 }
 
+type Controller interface {
+	Run(ctx context.Context) error
+	Shutdown()
+	GetPort() int
+}
+
+type ControllerStruct struct {
+	c Controller
+}
+
+func (c *ControllerStruct) StartServer() {
+	// this blocks
+	ctx := context.Background()
+
+	go func() {
+		if err := c.c.Run(ctx); err != nil {
+			return
+		}
+	}()
+}
+
+func (c *ControllerStruct) StopServer() {
+	c.c.Shutdown()
+}
+
+func NewControllerTest(c Controller) ControllerStruct {
+	ctrl := ControllerStruct{
+		c: c,
+	}
+
+	return ctrl
+}
+
+func (c *ControllerStruct) GetPort() int {
```

Comment:
Add a method for waiting instead of this one. That method would call c.c.GetPort() and pass the result to test.GetBaseURL(port). The url would be later passed to test.WaitTillServerReady(baseURL) in the same method.

Reply:
Make sure you use https://github.com/phayes/freeport for getting free ports, else port collisions and failed tests. Just fyi.